spring-integration-zookeeper/src/main/java/org/springframework/integration/zookeeper/leader/LeaderInitiator.java

```diff
@@ -0,0 +1,227 @@
+
+/**
+ * Bootstrap leadership {@link org.springframework.cloud.cluster.leader.Candidate candidates}
+ * with ZooKeeper/Curator. Upon construction, {@link #start} must be invoked to
+ * register the candidate for leadership election.
+ *
+ * @author Patrick Peralta
+ * @author Janne Valkealahti
+ * @author Gary Russell
+ *
+ */
+public class LeaderInitiator implements Lifecycle, InitializingBean, DisposableBean {
+
+	private static final String DEFAULT_NAMESPACE = "/spring-cloud/leader/";
+
+	/**
+	 * Curator client.
+	 */
+	private final CuratorFramework client;
+
+	/**
+	 * Candidate for leader election.
+	 */
+	private final Candidate candidate;
+
+	/**
+	 * Curator utility for selecting leaders.
+	 */
+	private volatile LeaderSelector leaderSelector;
+
+	/**
+	 * Flag that indicates whether the leadership election for
+	 * this {@link #candidate} is running.
+	 */
+	private volatile boolean running;
+
+	/** Base path in a zookeeper */
+	private final String namespace;
+
+	/** Leader event publisher if set */
+	private volatile LeaderEventPublisher leaderEventPublisher;
+
+	/**
+	 * Construct a {@link LeaderInitiator}.
+	 *
+	 * @param client     Curator client
+	 * @param candidate  leadership election candidate
+	 */
+	public LeaderInitiator(CuratorFramework client, Candidate candidate) {
+		this(client, candidate, DEFAULT_NAMESPACE);
+	}
+
+	/**
+	 * Construct a {@link LeaderInitiator}.
+	 *
+	 * @param client     Curator client
+	 * @param candidate  leadership election candidate
+	 * @param namespace  namespace base path in zookeeper
+	 */
+	public LeaderInitiator(CuratorFramework client, Candidate candidate, String namespace) {
+		this.client = client;
+		this.candidate = candidate;
+		this.namespace = namespace;
+	}
+
+	/**
+	 * Start the registration of the {@link #candidate} for leader election.
+	 */
+	@Override
+	public synchronized void start() {
+		if (!running) {
+			if (client.getState() != CuratorFrameworkState.STARTED) {
+				// we want to do curator start here because it needs to
+				// be started before leader selector and it gets a little
+				// complicated to control ordering via beans so that
+				// curator is fully started.
+				client.start();
+			}
+			leaderSelector = new LeaderSelector(client, buildLeaderPath(), new LeaderListener());
+			leaderSelector.setId(candidate.getId());
+			leaderSelector.autoRequeue();
+			leaderSelector.start();
+
+			running = true;
+		}
+	}
+
+	/**
+	 * Stop the registration of the {@link #candidate} for leader election.
+	 * If the candidate is currently leader, its leadership will be revoked.
+	 */
+	@Override
+	public synchronized void stop() {
+		if (running) {
+			leaderSelector.close();
+			running = false;
+		}
+	}
+
+	/**
+	 * @return true if leadership election for this {@link #candidate} is running
+	 */
+	@Override
+	public boolean isRunning() {
+		return running;
+	}
+
+	@Override
+	public void afterPropertiesSet() throws Exception {
+		start();
```

Comment:
This one doesn't look as good solution.
If you worry that `applicationContext` doesn't start your `Lifecycle`, just rework it to the `SmartLifecycle`.
Otherwise this implementation does not make sense to me...

Although I see that all these classes are copies from those in the SC. So, the fix must be done there after our M2 already. And we will come back to the dependency on the RC1 dead line